src/mpid/ch4/shm/posix/posix_init.h

```diff
@@ -4,231 +4,119 @@
-    mpi_errno = MPIDU_shm_barrier(MPIDI_POSIX_mem_region.barrier, num_local);
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_mpi_init_hook
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_POSIX_mpi_init_hook(int rank, int size, int *n_vnis_provided, int *tag_bits)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i;
 
-    if (mpi_errno)
-        MPIR_ERR_POP(mpi_errno);
+#ifdef MPL_USE_DBG_LOGGING
+    MPIDI_CH4_SHM_POSIX_GENERAL = MPL_dbg_class_alloc("SHM_POSIX", "shm_posix");
+#endif /* MPL_USE_DBG_LOGGING */
 
-    /* find our cell region */
-    MPIDI_POSIX_mem_region.Elements = cells_p[local_rank];
+    MPIDI_POSIX_global.am_buf_pool =
+        MPIDI_CH4U_create_buf_pool(MPIDI_POSIX_BUF_POOL_NUM, MPIDI_POSIX_BUF_POOL_SIZE);
 
-    /* Tables of pointers to shared memory Qs */
-    MPIR_CHKPMEM_MALLOC(MPIDI_POSIX_mem_region.FreeQ, MPIDI_POSIX_queue_ptr_t *,
-                        size * sizeof(MPIDI_POSIX_queue_ptr_t), mpi_errno, "FreeQ", MPL_MEM_SHM);
-    MPIR_CHKPMEM_MALLOC(MPIDI_POSIX_mem_region.RecvQ, MPIDI_POSIX_queue_ptr_t *,
-                        size * sizeof(MPIDI_POSIX_queue_ptr_t), mpi_errno, "RecvQ", MPL_MEM_SHM);
+    MPIR_CHKPMEM_DECL(3);
 
-    /* Init table entry for our Qs */
-    MPIDI_POSIX_mem_region.FreeQ[rank] = &free_queues_p[local_rank];
-    MPIDI_POSIX_mem_region.RecvQ[rank] = &recv_queues_p[local_rank];
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_INIT_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_INIT_HOOK);
 
-    /* Init our queues */
-    MPIDI_POSIX_queue_init(MPIDI_POSIX_mem_region.RecvQ[rank]);
-    MPIDI_POSIX_queue_init(MPIDI_POSIX_mem_region.FreeQ[rank]);
+    *n_vnis_provided = 1;
 
-    /* Init and enqueue our free cells */
-    for (i = 0; i < MPIDI_POSIX_NUM_CELLS; ++i) {
-        MPIDI_POSIX_cell_init(&(MPIDI_POSIX_mem_region.Elements[i]), rank);
-        MPIDI_POSIX_queue_enqueue(MPIDI_POSIX_mem_region.FreeQ[rank],
-                                  &(MPIDI_POSIX_mem_region.Elements[i]));
-    }
+    /* This is used to track messages that the eager submodule was not ready to send. */
+    MPIDI_POSIX_global.postponed_queue = NULL;
 
-    /* set route for local procs through shmem */
-    for (i = 0; i < num_local; i++) {
-        grank = local_procs[i];
-        MPIDI_POSIX_mem_region.FreeQ[grank] = &free_queues_p[i];
-        MPIDI_POSIX_mem_region.RecvQ[grank] = &recv_queues_p[i];
+    MPIR_CHKPMEM_MALLOC(MPIDI_POSIX_global.active_rreq,
+                        MPIR_Request **,
+                        size * sizeof(MPIR_Request *), mpi_errno, "active recv req", MPL_MEM_SHM);
 
-        MPIR_Assert(MPIDI_POSIX_ALIGNED
-                    (MPIDI_POSIX_mem_region.FreeQ[grank], MPIDI_POSIX_CACHE_LINE_LEN));
-        MPIR_Assert(MPIDI_POSIX_ALIGNED
-                    (MPIDI_POSIX_mem_region.RecvQ[grank], MPIDI_POSIX_CACHE_LINE_LEN));
+    for (i = 0; i < size; i++) {
+        MPIDI_POSIX_global.active_rreq[i] = NULL;
```

Comment:
I'm not sure what you mean. Aren't _all_ processes shm processes?

Reply:
I'm confused. I don't think that's what `active_rreq` does. This is an O(p) array of requests that need to be progressed.

Reply:
Yes, it is used in progress. But it's indexed by global rank. The reason I guess it uses global rank instead of shm_rank is to saving translation overhead.

Reply:
Ok. I see what you mean.

In that case, I support the idea that we add this optimization later.